openpype/hosts/maya/plugins/publish/validate_light.py

```diff
@@ -0,0 +1,111 @@
+import json
+import os
+
+from maya import cmds
+
+import pyblish.api
+
+from openpype.pipeline.publish import (
+    ValidateContentsOrder, PublishValidationError
+)
+from openpype.settings.lib import convert_to_int_or_float
+
+
+class ValidateLightRequired(pyblish.api.InstancePlugin):
+    """Validate Light required settings."""
+
+    order = ValidateContentsOrder
+    hosts = ["maya"]
+    families = ["light"]
+    label = "Validate Light Required"
+    light_types = []
+    attribute_values = []
+
+    def process(self, instance):
+        # Collect validation data.
+        invalid_nodes = []
+        attribute_nodes = instance.data["setMembers"].copy()
+        file_nodes = []
+        for node in instance.data["setMembers"]:
+            shapes = cmds.listRelatives(node, fullPath=True, shapes=True)
+            attribute_nodes.extend(shapes)
+
+            for shape in shapes:
+                file_nodes.extend(
+                    cmds.listConnections(
+                        shape,
+                        skipConversionNodes=True,
+                        type="file",
+                        destination=False
+                    )
+                )
+
+            shape_types = {x: cmds.nodeType(x) for x in shapes}
+            invalid_types = set(shape_types.values()) - set(self.light_types)
+            invalid_nodes.extend(
+                [x for x in shape_types if shape_types[x] in invalid_types]
+            )
```

Comment:
This making of a dict would've made much more sense if instead the dict lookup would be reversed since you're basically not using the key mapping to your benefit at all for the lookup requiring the re-iteration of all the shapes again.

```python
shapes_by_type = defaultdict(list)
for shape in shapes:
    nodetype = cmds.nodeType(shape)
    shapes_by_type[nodetype].append(shape)

invalid_types = set()
for nodetype, nodetype_shapes in shapes_by_types.items():
    invalid_types.add(nodetype)
    invalid_nodes.extend(nodetype_shapes)
```

But also since the collected invalid types don't appear to be logged further down this could've also been as simple as:
```python
for shape in shapes:
    if cmds.nodeType(shape) not in self.light_types:
        invalid_nodes.append(shape)
```